### ISSUE TITLE
semantic highlighting not being attempted on large files?

### ISSUE THREAD

**Issue by seebs** [golang/vscode-go#2601]
### What version of Go, VS Code & VS Code Go extension are you using?

1.19, 1.68.0 (mac), 0.37.0

### Share the Go related settings you have added/edited

I think just turning on semantic highlighting.

### Describe the bug

To be very clear: This is **NOT** the thing where gopls disables semantic highlighting on large files.

By default, gopls is shipped configured to refuse semantic highlighting on files over 100,000 bytes (maxFullFileSize in internal/lsp/semantic.go). I am aware of this. I have rebuilt gopls with a different limit.

Prior to today, this worked; my rebuilt gopls worked and gave semantic information for larger files too.

Today, I upgraded MacOS. (that can't possibly matter, right? hahaha. cries into beer.) When I restarted code, it was no longer highlighting large files. So I rebuilt gopls in case that was the problem. Still not highlighting large files.

So I got curious, and added logging to gopls. Right before the check against the file size, I log the output, so I get messages like this:

```
semantic.go:100: considering semantics for file, len 12952
```

And what I've found is that I do not get these messages when the file size is 100KB or more.

This suggests to me that the extension isn't *trying* to request the highlighting. So, yes, I'm aware that gopls by default does not allow this. I changed it. So I'm assuming that the extension is helpfully, being aware of that, not sending the requests when it expects them to be refused, maybe? But that is in my case unhelpful, and I can't find a configuration knob or setting for it.

### Steps to reproduce the behavior:

1. Turn on semantic highlighting.
2. Make a go file over 100KB in size.
3. Modify your copy of gopls to allow semantic highlighting of large files.
4. Nothing Happens.

**Comment by pjweinb** [comment]
the log file would show if the client is asking for semantic tokens.

**Comment by seebs** [comment]
Which log file, and how do I look at it?

I added logging to my copy of `gopls` that showed activity when opening small files in vscode, but not when opening large files, but I don't know whether that means that the client wasn't making a request to gopls at all, or gopls was refusing the request earlier in code i haven't read yet.

Looking at the source for the vscode-go extension, i can't figure out at all when/where it is issuing the requests -- I can't find anything other than the UI/config stuff that even contains the string `semanticTokens` or other similar variants. (Disclaimer: I don't really know JavaScript and I don't know anything about the plugin architecture so I don't know what I'd be looking for.)

**Comment by seebs** [comment]
Huh. There's an interesting difference. If I open a new tab for one of the affected (large) files, I get the messages from gopls indicating it thought it was doing semantic tokens for them, but only when it's a new tab. If I change to a tab that's already open, small files will always show that message, but large files won't. Doesn't result in actually displaying any highlights, though.

**Comment by pjweinb** [comment]
and which version of vscode (or vscode-insiders) are you using?

**Comment by seebs** [comment]
```

Version: 1.68.0 (Universal)
Commit: 4af164ea3a06f701fe3e89a2bcbb421d2026b68f
Date: 2022-06-08T11:44:00.250Z (7 mos ago)
Electron: 17.4.7
Chromium: 98.0.4758.141
Node.js: 16.13.0
V8: 9.8.177.13-electron.0
OS: Darwin x64 22.1.0
```

"7 months ago" sounds really old and sort of weird to me. I'll see whether I can update.

**Comment by seebs** [comment]
Okay, updated, now 1.74.2. (Apparently, the instructions for getting around the "read-only volume" thing don't work for me, but I could just download a new one and install it.) No change in any of the other behavior, though.

**Comment by pjweinb** [comment]
Here's something to try: Instead of adjusting the file size, force vscode
to only use range semantic token queries:
At the end of lsp/workspace.go, ih the semanticTokensRegistration()
function, change the Full: setting to
Full: false,

If you do try that, let me know if it gives the behavior you want, both on
big and on small files.

**Comment by seebs** [comment]
Hmm. The only change is there's one fewer attempt each time, so on small files I get logs suggesting two queries, and on the large file, I see only one. No formatting on the large file.

**Comment by seebs** [comment]
Update, I poked at this a bit again recently. Newer vscode, newer gopls, etcetera.

Still the same behavior; no matter what setttings I make or change in gopls, the extension never actually gets semantic token information for files over 100,000 bytes. This includes the suggested "disable full, only allow ranges" behavior, setting the cap to 1M bytes, and so on.

As noted above, if I add logging to gopls, it looks like it's getting multiple requests, but on large files, it gets one fewer request, so it looks very much like something in `vscode-go` is choosing not to send the additional request, rather than like `gopls` isn't honoring it when it gets the request.

**Comment by hyangah** [comment]
@pjweinb @seebs Is there any opensource project that we can use to test?

"This suggests to me that the extension isn't trying to request the highlighting" --> if this is the case, this can be a bug in vscode. Is this problem reproducible in other editors?

**Comment by seebs** [comment]
So I found the problem, and it wasn't vscode.

It's that the one extension I use which uses semantic highlighting ALSO has a file size limit, and its file size limit is very close to the default file size limit for go's LSP semantic highlighting. And this file size limit feature appears to have been added recently.

it was https://marketplace.visualstudio.com/items?itemName=MatthewNespor.vscode-color-identifiers-mode all along.

This also probably explains why I was seeing multiple requests made to the LSP -- one from whatever else vscode was doing, which was working, and another from this extension, which was stopping when the file size was larger than its default setting.

So nevermind, vscode is actually fine, it's just this extension adding a setting I didn't know about since I first installed it. Sorry for the trouble!
